corehq/apps/userreports/decorators.py

```diff
@@ -39,3 +41,32 @@ def _inner(*args, **kwargs):
         except TableNotFoundException:
             raise TableNotFoundWarning
     return _inner
+
+
+def ucr_context_cache(vary_on=()):
+    """
+    Decorator which caches calculations performed during a UCR EvaluationContext
+    The decorated function or method must have a parameter called 'context'
+    which will be used by this decorator to store the cache.
+    """
+    def decorator(fn):
+        assert 'context' in fn.func_code.co_varnames
+        assert isinstance(vary_on, (tuple, list))
+
+        @wraps(fn)
+        def _inner(*args, **kwargs):
+            # shamelessly stolen from quickcache
+            callargs = inspect.getcallargs(fn, *args, **kwargs)
+            context = callargs['context']
+            prefix = '{}.{}'.format(
+                fn.__name__[:40] + (fn.__name__[40:] and '..'),
```

Comment:
It'd be nice to be able to override this. I could see some instances where we want to share a cache (especially in custom expressions)

Reply:
I was thinking about that as well, but I figured if we'd want to share a cache, then it'd probably also want the same implementation.  That's basically what I did here: https://github.com/dimagi/commcare-hq/pull/18778/files

Right now this only works with simple arguments, but if/when it's expanded to support complex stuff, that could definitely be useful, and I'd support adding it.
Consider two expressions which get data from a form - one starts off with a form object, and the other starts off with only a form ID.  They'd both want to share a key, but we wouldn't want the first scenario to re-fetch the form, so the implementations would differ.  For the first scenario to work, though, this decorator would have to support something like `'form.get_id'` as a key.  I think I'll punt on that until I need it.

Reply:
that makes sense 👍